### ISSUE TITLE
Feature request: Open with external diff tool

### ISSUE THREAD

**Issue by rlabrecque** [desktop/desktop#1765]
This is something that I find doing a lot in my everyday workflow and it's a pretty standard option in most vcs clients I've used.

This is somewhat related to #1554

#### Steps:

- [ ] Add 'External Diff' to Changes -> File context menu
- [ ] Add 'External Diff' to Repo list -> Repo context menu
- [ ] Add 'External Diff' to History -> Commit context menu
- [ ] Add 'External Diff' to History -> Commit -> File context menu
- [ ] Allow the user to specify their diff tool in the Advanced Options.
- [ ] If no diff tool is specified don't show the field, or show the option and bring them to the Advanced Options to set it up?

Thoughts?

**Comment by joshaber** [comment]
Is this for resolving merge conflicts, or purely as another way of viewing diffs?

**Comment by rlabrecque** [comment]
This was just another way of viewing diffs, using dedicated diffing tools (Beyond Compare, AraxisMerge, etc). One example of this is use case is diffing Unreal Engine uasset's. In something like perforce (P4V) you add an override to this command based on file extension, so that you can jump right from version control to a diff with no extra steps. It's great!

But, close this for now and mark it as future-proposal because it's definitely out of the scope for 1.0.

**Comment by joshaber** [comment]
👍 💖

**Comment by ashleydavis** [comment]
+1

**Comment by alexiswilliams** [comment]
This would be great. This is something I used for years with that "other" tool, Perforce. P4V Merge was a great tool to use when reviewing changes in code by other developers or myself to make changes in the review and annotate.

**Comment by aking7** [comment]
I would also love this, so I can diff MATLAB binary files, as in:

https://se.mathworks.com/help/matlab/matlab_prog/customize-external-source-control-to-use-matlab-for-comparison-and-merge.html

**Comment by dmkash** [comment]
I would love to see this implemented so that I can use [Kaleidoscope](https://www.kaleidoscopeapp.com/) as the diff/merge tool.

**Comment by ashleydavis** [comment]
Yeah GitHub Desktop is unusable without external diff. 

Personally I use Winmerge and Beyond Compare. I've actually bought Beyond Compare because it is so good a resolving complicated merges. Traditional version control software allows for external diff and merge.  

There was a period where SourceTree was very buggy and slow (and Tortoise authentication was a pain) so I was considering GitHub Desktop. Now that SourceTree is back on form there is no comparison.

**Comment by TtamNoslot** [comment]
I want this too for Beyond Compare integration with GitHub Desktop.

**Comment by aka-emi** [comment]
Having all github's code coloring abilities in one single difftool would be nothing but wonderful.

**Comment by Ziflin** [comment]
Is there an open task for this? I was about to switch to GitHub for Windows from SourceTree and realized there's no right-click / External Diff tool option.

**Comment by rsiw2000** [comment]
It would be great if I could use an external diff program (such as Beyond Compare, KDiff, Meld, Winmerge and so on) to view diffs and resolve merge conflicts.

**Comment by UrviAhuja** [comment]
Would love to have this

**Comment by sal48** [comment]
@joshaber Any updates on this? It was closed with a future proposal tag, but we don't know how to follow from there.

**Comment by williamsandonz** [comment]
I would love this feature too! As I payed for a Beyond Compare license and like to use it for merges.

**Comment by tfgstudios** [comment]
Why is this closed? Will it be implemented or not?

**Comment by Sauraus** [comment]
Same question why was this closed? Its clearly not resolved.

**Comment by tierninho** [comment]
Please note that issues marked as `future-proposal` are not forgotten. We aim to circle back once we determine there is room in the roadmap to implement such a feature. As of now, we have no definite update.

In the meantime, we welcome any constructive feedback on how to make this proposed feature even better. Thanks for your interest and for also using Desktop!

**Comment by Sauraus** [comment]
Will you accept pull requests that implement this? Even if this will be limited to a specific diff tool to begin with?

**Comment by tierninho** [comment]
@Sauraus The issue is not yet labeled as `help-wanted` however I am cc'ing @billygriffin for thoughts on updating this.

**Comment by Sauraus** [comment]
> @Sauraus The issue is not yet labeled as `help-wanted` however I am cc'ing @billygriffin for thoughts on updating this.

Not sure when and if I'd get around to doing it, but seeing as the issue has been open for two years? :)

**Comment by nakott** [comment]
The way GitHub Desktop for Windows is not helpful for me. For me, the best tool to diff files is Beyond Compare. So, calling an external diff tool is very important, otherwise I cannot use GitHub Desktop.

**Comment by nakott** [comment]
The way GitHub Desktop for Windows shows diffs is not helpful for me. For me, the best tool to diff files is Beyond Compare. So, calling an external diff tool is very important, otherwise I cannot use GitHub Desktop.

**Comment by sitaga** [comment]
+1  I am new to GitHub, from Subversion - not being able to do a diff between branches seems like epic fail

**Comment by robertcatgithub** [comment]
wtf still not available ???!!! Why is this closed ?

**Comment by ashleydavis** [comment]
Yeah it's crazy.

Just use SourceTree (it isn't as buggy as it used to be).

https://www.sourcetreeapp.com/

**Comment by moneythumb** [comment]
I would think something like this would have been addressed, given that its now 3 years later.
What does it take to make this happen?

**Comment by xiaocongli01** [comment]
any good news?

**Comment by jlczuk** [comment]
So this was closed in 2017 with a future proposal label because it would not make 1.0.  Here we are in 2021 with version 2.6.1 and still no support to use an external diff tool.

This could be accomplished in a slightly unpolished way by having settings to specify the command-line and options for both the external diff and merge commands to use, much the way one uses git config.

Having canned selections for the most popular diff/merge tools would be nice, but it isn't essential.

@joshaber @rlabrecque

**Comment by jsterner-ss** [comment]
Really want this feature

**Comment by aklef** [comment]
@niik @sergiou87 @tidy-dev @billygriffin your recent work in this repo has been wonderful! Bravo and good health to you all. 

Can one of you please tell @jlczuk @rlabrecque and myself when we can reasonably expect this feature in github desktop? I'll be thrilled just to have desktop read my gitconfig and launch whatever is set in there as difftool

**Comment by rlabrecque** [comment]
For what it's worth I never really ended up using GHD as more powerful alternatives started coming out. It's clear GHD is targeted at a different audience than most of us, and that's fine. But I don't particularly care about this issue myself anymore 👍

**Comment by ashleydavis** [comment]
I'm still using SourceTree. It has all the professional features but performance isn't the best.

What are other devs using these days?

**Comment by ashleydavis** [comment]
I've asked this question on Twitter: 
https://twitter.com/ashleydavis75/status/1357815598902779904

If you are on there, please jump in and tell me what you are using!

**Comment by aklef** [comment]
@rlabrecque that makes sense. I have continued using desktop for its simplicity and ease of use and this one feature is the only thing in the last few years I've felt was lacking from the tool, hence my comment. desktop is wonderful for 99% of my work

**Comment by jen-ren** [comment]
Such a bummer that external diff is not supported ....

**Comment by Thaval** [comment]
Created almost 4 years ago.. is it really that hard to implement such a simple feature? Aborting github desktop and switching to source tree again. Without external diff, this is pretty useful. As if I want to open Visual Studio Code or Visual Studio merging... wtf :D

**Comment by kfredericks-ip** [comment]
Please implement external diff support. It would be much appreciated. Thank you.

**Comment by rsiw2000** [comment]
It's very clear that Github Desktop is running to cover the most common  features ... while this approach take users away from it ....

**Comment by Thaval** [comment]
Are you willing to implement it? I'm just curious as the request is 4 years old and the feature seems pretty trivial for me (let user specify a path and let him set some custom parameters).

**Comment by germain-italic** [comment]
please enjoy my upvote on this request

**Comment by billygriffin** [comment]
Hi folks, despite the engagement in this issue, this actually isn't a feature we see requested that frequently so it hasn't risen to the level of us working on it on our roadmap. I don't think anyone is against this, we would just need to dig into it further and haven't had the opportunity to do that yet. I can say there's a high probability this won't be shipped in the next 3-6 months, and beyond that all I can say is that we'll continue to consider it alongside other possible features, all with people who would really like to see them in Desktop. Sorry that doesn't give you a better answer, and I wish I had one.

If you'd like to see this feature happen, please use the 👍 reaction on the issue, as we do use that as an input into our prioritization of features (though it's not the only consideration). Comments like "I want this too" aren't helpful and cause noise in people's notifications so please try to avoid those. Thanks!

**Comment by akochepasov** [comment]
Beyond Compare is a tool I use a lot during development and code reviews. Please consider adding the diff-merge customization.

**Comment by Chimera78** [comment]
I know comments don't help. But even "ancient" WinCVS allows the usage of external diff tools. So please offer this feature to all the users trying to move to modern tools such as GIT and use a decent UI such as GitHub Desktop. That would just increase acceptance from us "old dinosaurs" trying to catch up with new modern tools.

**Comment by arnadan** [comment]
So the reason for not working on this issue is that it is not requested often enough? So I need to add this comment just to let you know that I would realy appreciate this feature! Have switched from SVN to github recently and am really missing the option to include Beyond Compare in github Desktop.

**Comment by jbelina-godaddy** [comment]
+1 requesting diff-merge customization so that I can use Beyond Compare.

**Comment by ydubov** [comment]
+1

**Comment by avigailIntel** [comment]
+1

**Comment by iysabbah** [comment]
+1

**Comment by Thaval** [comment]
+1

**Comment by KASDPaul** [comment]
+1

**Comment by nakott** [comment]
+1

**Comment by derekvantonder** [comment]
This is a really basic feature for any Git tool to have. Like, literally every other desktop Git tool has this, it's so simple to implement, just give users command-line options with a couple of tokens.

**Comment by feitla** [comment]
Yes, can we please get this feature. I've already changed the default program to VS Code on all those file times but it still suggests XCode.

**Comment by KarthikAbiram** [comment]
This would be something that would make a **BIG difference** to those who work with code which use custom binary files like LabVIEW. 

**@All, Those who would like this feature to be implmented, please add a like to the 1st post and add a line of comment on why you need this feature - This would help making this feature a priority.**

**Comment by pcipera** [comment]
+1

**Comment by rg2024** [comment]
+1

**Comment by pierrebestwork** [comment]
"Beyond Compare is a tool I use a lot during development and code reviews. Please consider adding the diff-merge customization"

Ditto

**Comment by ashleydavis** [comment]
Recently I've bought Fork, which is a Git client from an indie dev.

It's really nice and it supports external diff tools.

I'm not affiliated with the developer of Fork in any way, but I can wholeheartedly recommend it. 

There's a generous free trial, so please try it, you won't be disappointed. 

https://git-fork.com/

**Comment by SomeEngineer42** [comment]
This feature is preventing us from adopting this tool at an enterprise level. We often have to diff much more complicated files that people use tools like BeyondCompare for (xlsx files, etc) and missing support makes this a non-starter

**Comment by ZZ-Zhou** [comment]
this feature suppose to be a common and standard feature, I don't understand why it is so hard to be implemented in Github desktop?  This definitely prevented me to use github desktop.

**Comment by vishnuprathish** [comment]
This feature is a must have for doing better comparisons for several external file types. Jupyter notebooks, certain types of binaries etc.

**Comment by phaselden** [comment]
Agreed. This is really the only piece of functionality which is preventing Desktop from becoming the default tool

**Comment by JeffSagerEnchoice** [comment]
This is keeping me from using your tool, and it is such a showstopper for me that I suffer through using SourceTree!
I have written many internal tools for our application that open an external tool, the code is no more than a few lines.  I do realize the bigger job is adding a place to configure this, but you could just read my .gitconfig file and use the data from there!

**Comment by ashleydavis** [comment]
I recommend adopting Fork for those still on Sourcetree. You won't regret changing.

**Comment by fbrioccia** [comment]
+1

**Comment by mbyt** [comment]
+1 yes, this is a crucial feature when working with e.g. jupyter notebooks, fodt files, Max/MSP maxpats... It is generally needed for all file formats which contain things like output sections, information based on timestamps, non deterministic orders of dictionaries.

For diffing those files, I currently need to use another git tool in parallel to Github desktop for viewing the diffs. It would be nice and smoother, if I could double click the file inside Github desktop and then see the diffs for those filetypes in a custom diff tool.

**Comment by bjudkewitz** [comment]
> Agreed. This is really the only piece of functionality which is preventing Desktop from becoming the default tool

Fully agree

**Comment by xumix** [comment]
>But, close this for now and mark it as future-proposal because it's definitely out of the scope for 1.0.

@rlabrecque @joshaber 5 years passed, v1.0 is long past due
Could you please reopen this?

**Comment by AnsisMalins** [comment]
Bump.

**Comment by aukgit** [comment]
+1

**Comment by brian-insitro** [comment]
An emphatic +1... There are diff features that I sorely need from an external editor.

**Comment by Patel-Aakash** [comment]
+1

**Comment by aukgit** [comment]
+1

**Comment by samirbavkar** [comment]
yes, adding my vote. Please allow support of external diff tools.

**Comment by ksultzz** [comment]
+1 for Beyond Compare and other external diff tools

**Comment by alvinashcraft** [comment]
Adding my vote to have the ability to launch external diff tools for diff and merge operations.

**Comment by ddbrown30** [comment]
Wow. More than 6 years since this first came up. This is clearly not a priority that the dev team agrees with.

**Comment by derekvantonder** [comment]
> Wow. More than 6 years since this first came up. This is clearly not a priority that the dev team agrees with.

It's cool, if I need to merge/diff I don't care because I just open the directory in JetBrains RIDER or any other IntelliJ tool and they have far better Git tools than anything in GitHub Desktop. GitHub Desktop remains my go-to for quickly cloning projects, anything requiring merging or diff'ing happens in RIDER. I have a sneaking suspicion the GitHub desktop devs do this too 😂

**Comment by ashleydavis** [comment]
Just use Fork it's so much better.

https://git-fork.com/

(I'm an enthusiastic user)

**Comment by Jamsers** [comment]
Yet another user requesting this simple feature. Would go a long way towards making GitHub Desktop much more usable.

**Comment by liviupopoviciaxiamed** [comment]
I arrived here to see how to configure a diff tool in github desktop and i found you actually can't. pretty sad... i will search for another tool for git...

**Comment by andrewrnewton** [comment]
Here to add my +1 for this too. Github desktop makes git bareable (because it's generally lousy). But when there are large code changes, the built in diff viewer is not capable. If I thought it was worth £80 on a licence for Beyond Compare that must mean that most of the free stuff doesn't cut it! 

After writing software for 20 years, I feel the most basic features of diffing should be side by side with modifiable code alignment. But equally, code is more complex than it was in the late 90s and we are doing more and more with it. The tools need to keep up with that complexity too.

**Comment by andrewrnewton** [comment]
> Just use Fork it's so much better.
> 
> https://git-fork.com/
> 
> (I'm an enthusiastic user)

Whilst I wouldn't begrudge them 50 quid for their product which looks nice, it doesn't look like you can add the diff tool I already paid 80 quid for to that either...

**Comment by ashleydavis** [comment]
> > Just use Fork it's so much better.
> > https://git-fork.com/
> > (I'm an enthusiastic user)
> 
> Whilst I wouldn't begrudge them 50 quid for their product which looks nice, it doesn't look like you can add the diff tool I already paid 80 quid for to that either...

You can definitely add external diff and merge tools to fork.

**Comment by MCR-TEA** [comment]
I **CANNOT** Review any Check-In without BeyondCompare!

### Justification:
There are many times when Files appear the same, but it's a change in the Hidden Hex Values (like a NULL Code),
 &nbsp; &nbsp; or the Encoding's changed (which BeyondCompare will show me in its comparison).

Plus, BeyondCompare has so many useful features to:
 &nbsp; &nbsp; 1.) View all the Code  - OR -  Show only Differences.
 &nbsp; &nbsp; 2.) Option to Ignore Unimportant Differences.
 &nbsp; &nbsp; 3.) Edit the File within the Diff-Tool (keep some changes, ignore others, or Manually edit lines).
 &nbsp; &nbsp; 4.) Search.
 &nbsp; &nbsp; 5.) Show Whitespace.
 &nbsp; &nbsp; 6.) Align to a Different Line where you prefer the Matching/Comparison to Continue from.
 &nbsp; &nbsp; 7.) Table Comparisons!
 &nbsp; &nbsp; 8.) View Encoding.
 &nbsp; &nbsp; 9.) View the actual Encoded Character-Based Hexadecimal Values.
 &nbsp; 10.) etc...

It is an Invaluable Tool for code development and if you're ONLY using GitHub Desktop to check-in your files,
 &nbsp; &nbsp; then, my friend, you're doing it wrong.

### Workaround:
In conjunction with GitHub-Desktop, I also use **TortoiseGit**.
TortoiseGit allows me to specify an External-Diff App for 2-way _and_ 3-way merges (I rarely have 3-way's though).

I just Right-Click on my Folder (or File) in Windows Explorer and Select "_TortoiseGit_"->"_Diff_".
From there it shows me the List of Files that have changed and,
 &nbsp; &nbsp; just by Double-Clicking on them, I may View the Diffs in BeyondCompare.

### Bonus:
What I also like about this Diff-View in TortoiseGit is how it shows me a Detailed-Pane of:
 &nbsp; &nbsp; Path, FileName, Extension, Status, Lines Added, Lines Removed, Last Modified, and Size.
 &nbsp; &nbsp; Note: All of these Detailed-Columns are Configurable to Show/Hide _and_ Move around.
I may then Sort by any of these or use them to look for specific FileTypes (or Edits) I did in a specific moment in Time.
This is yet another basic (yet extremely powerful) feature that GitHub-Desktop does _**not**_ have - and it's a shame it doesn't.

I may also Right-Click on my Folder (or File) in Windows Explorer and Select "_TortoiseGit_"->"_Show Log_".
From here I may view **_all_** of our past Check-Ins and Compare/Diff those to any Check-In I choose (or against my Working Copy).

TortoiseGit also has other Features you may find useful, but GitHub-Desktop is just so much easier and intuitive when it comes time to Pulling Down a Branch or Checking-In Code.  _These 2 Apps are like my Tag-Team for Check-Ins_.

Hope this helps and sorry for offending any based-devs who don't need External-Diff's for Troubleshooting and Sanity-Checks.
I'm not a command-line warrior, so I appreciate these GUI's that make my life easier when it comes to code management.
### Example of TortoiseGit:
![image](https://github.com/desktop/desktop/assets/150230844/2d21ded6-e664-45e9-94c5-4420b3eb384b)
### Example of BeyondCompare:
![image](https://github.com/desktop/desktop/assets/150230844/b82be0c2-5c99-44ce-a9ad-2d9b09974f8a)

**Comment by XavierMurrow-ui** [comment]
+1. I am yet another weary traveller, trying to find a solution to this horrible issue. I plead for this feature, nearly 7 years after the original comment.

**Comment by Tommuh** [comment]
+1

**Comment by tiagokeller-fortis** [comment]
+1

**Comment by tushar-dev-sharma** [comment]
+1 - we work with binary formats and would love to be able to diff them with an external tool.

**Comment by TheMindAtWork** [comment]
Another up vote to please add this feature!
Beyond Compare is one of the best diff tools and I suffer without it in GitHub Desktop.

**Comment by BerlinoPotato** [comment]
+1 Beyond Compare will be great. Especially if we can select different commits to compare to current one too.

**Comment by jacob-stephenson-st** [comment]
+1, I cannot use GitHub Desktop until this is implemented.

**Comment by gabriel-gn** [comment]
Can this be opened again? Since we now have implementation for custom editors with args, and custom shells on the app

**Comment by shrddr** [comment]
[iAmWillShepherd](https://github.com/iAmWillShepherd) added this to the [1.1](https://github.com/desktop/desktop/milestone/11) milestone [on Oct 18, 2017](https://github.com/desktop/desktop/issues/1765#event-1299320732)

![image](https://github.com/user-attachments/assets/f1b97eea-c52d-4b06-8735-8cb5569fdce4)

**Comment by kylefox** [comment]
I'd love to be able to use Kaleidoscope with the GitHub app. Quite surprised this isn't possible 🤔

**Comment by maitrey** [comment]
I would like support for this feature, Could this feature be reopened please?

**Comment by JoHoGH** [comment]
Another request to reopen this please. I avoid GitHub desktop because of this and use tortoisegit instead (configured to use WinMerge for excel files), but my organisation is enforcing it.

Just realized I can use "Repository > Open in Git Bash" from within github desktop  & then " git difftool -d" (with Winmerge configured). Almost as quick as having a button avaialble.

**Comment by Harkirat444** [comment]
Please reopen this. Should be a standard feature in Github.
